src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClientX/PoolingDataSource.cs

```diff
@@ -0,0 +1,81 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#if NET8_0_OR_GREATER
+
+using System;
+using System.Collections.Concurrent;
+using System.Net.Security;
+using System.Security.Cryptography.X509Certificates;
+using System.Threading;
+using System.Threading.Channels;
+using System.Threading.Tasks;
+using Microsoft.Data.ProviderBase;
+using Microsoft.Data.SqlClient;
+
+namespace Microsoft.Data.SqlClientX
+{
+    /// <summary>
+    /// A pooling implementation of SqlDataSource. Connections are recycled upon return to be reused by another operation.
+    /// </summary>
+    internal sealed class PoolingDataSource : SqlDataSource
+    {
+        private DbConnectionPoolGroupOptions _connectionPoolGroupOptions;
+
+        internal int MinPoolSize => _connectionPoolGroupOptions.MinPoolSize;
+        internal int MaxPoolSize => _connectionPoolGroupOptions.MaxPoolSize;
+
+        /// <summary>
+        /// Initializes a new PoolingDataSource.
+        /// </summary>
+        //TODO: support auth contexts and provider info
+        PoolingDataSource(SqlConnectionStringBuilder connectionStringBuilder,
+            SqlCredential credential,
+            RemoteCertificateValidationCallback userCertificateValidationCallback,
```

Comment:
Why do these callbacks exist?

Reply:
I think this is a holdover from npgsql inspiration. I can clean them up and come back around later as the auth flows are fleshed out.